plugin-api/src/main/java/org/hyperledger/besu/plugin/services/storage/KeyValueStorage.java

```diff
@@ -61,15 +63,23 @@
   Optional<byte[]> get(byte[] key) throws StorageException;
 
   /**
-   * Performs an evaluation against each key in the store, keeping the entries that pass, removing
-   * those that fail.
+   * Returns a stream of all keys.
    *
-   * @param retainCondition predicate to evaluate each key against, unless the result is {@code
-   *     null}, both the key and associated value must be removed.
-   * @return the number of keys removed.
-   * @throws StorageException problem encountered when removing data.
+   * @return A stream of all keys in storage.
+   * @throws StorageException problem encountered during the retrieval attempt.
+   */
+  Stream<byte[]> streamKeys() throws StorageException;
+
+  /**
+   * Attempt to delete the value corresponding to the given key. For some storage mediums we want to
+   * be able to abandon the delete under certain circumstances. For example, to abandon the attempt
+   * if we can't acquire a lock on the underlying storage.
```

Comment:
We probably need to make the requirements for this method clearer - is it allowed to wait for the lock to be available or must it always immediately fail if the lock is unavailable?  Otherwise we risk pruning not working or getting deadlocks with some storage plugins because they don't exhibit the behaviour we need here.